execution/trigger/multitrigger.go

```diff
@@ -43,39 +49,44 @@ func (m *MultiTrigger) UpdateWatermark(ctx context.Context, tx storage.StateTran
 	for i := range m.triggers {
 		err := m.triggers[i].UpdateWatermark(ctx, tx.WithPrefix(m.prefixes[i]), watermark)
 		if err != nil {
-			return errors.Wrapf(err, "couldn't update watermark in trigger with index %d: %v", i, m.triggers[i])
+			return errors.Wrapf(err, "couldn't update watermark in Trigger with index %d: %v", i, m.triggers[i])
 		}
 	}
 
 	return nil
 }
 
-func (m *MultiTrigger) PollKeyToFire(ctx context.Context, tx storage.StateTransaction) (octosql.Value, error) {
+func (m *MultiTrigger) PollKeysToFire(ctx context.Context, tx storage.StateTransaction) ([]octosql.Value, error) {
+	var outKeys []octosql.Value
 	for i := range m.triggers {
-		key, err := m.triggers[i].PollKeyToFire(ctx, tx.WithPrefix(m.prefixes[i]))
-		if err == nil {
+		keys, err := m.triggers[i].PollKeysToFire(ctx, tx.WithPrefix(m.prefixes[i]))
+		if err == nil && len(keys) > 0 {
 			for j := range m.triggers {
 				if j != i {
-					err := m.triggers[j].KeyFired(ctx, tx.WithPrefix(m.prefixes[j]), key)
+					err := m.triggers[j].KeysFired(ctx, tx.WithPrefix(m.prefixes[j]), keys)
 					if err != nil {
-						return octosql.ZeroValue(), errors.Wrapf(err, "couldn't mark key fired in trigger with index %d after trigger with index %d fired", j, i)
+						return nil, errors.Wrapf(err, "couldn't mark keys fired in Trigger with index %d after Trigger with index %d fired", j, i)
 					}
 				}
 			}
-			return key, nil
-		} else if err != execution.ErrNoKeyToFire {
-			return octosql.ZeroValue(), errors.Wrapf(err, "couldn't poll key to fire in trigger with index %d: %v", i, m.triggers[i])
+			if outKeys == nil {
```

Comment:
I think you could just declare outKeys as
`
make([]octosql.Value, 0)
`

and then get rid of this if statement

Reply:
nil and this make are equal. This is just a small premature optimisation. I think without append this will be faster.

Reply:
No unnecessary copying.